Aplib.Core/Action.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System;
+
+namespace Aplib.Core
+{
+    /// <summary>
+    /// Describes an action that can be executed and guarded.
+    /// </summary>
+    /// <typeparam name="TQuery">The type of the query of the action</typeparam>
+    public class Action<TQuery>
+    {
+        private TQuery? _storedResult;
+
+        /// <summary>
+        /// Gets or sets the effect of the action.
+        /// </summary>
+        public System.Action<TQuery> Effect { private get; set; }
+
+        /// <summary>
+        /// Gets a value indicating whether the action is actionable.
+        /// </summary>
+        public bool Actionable => Query!.Invoke() is not (false or null);
+
+        /// <summary>
+        /// Gets or sets the query of the action. Can return a value, which is stored and can be used in the effect.
+        /// </summary>
+        /// <remarks>A boolean value can also be used. If an object is used, use null to mark the action as unactionable.</remarks>
+        public Func<TQuery> Query { private get; set; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Action{TQuery}"/> class.
+        /// </summary>
+        /// <param name="effect">The effect of the action.</param>
+        /// <param name="query">THe query of the action.</param>
+        public Action(System.Action<TQuery> effect, Func<TQuery> query)
```

Comment:
As described above, I think the query should be nullable, and have default value null (meaning the action will always be performed when the corresponding goal is active)